packages-content-model/roosterjs-content-model-plugins/lib/edit/tabUtils/handleTabOnParagraph.ts

```diff
@@ -0,0 +1,75 @@
+import { createSelectionMarker, createText } from 'roosterjs-content-model-dom';
+import { setModelIndentation } from 'roosterjs-content-model-api';
+import type {
+    ContentModelDocument,
+    ContentModelParagraph,
+    RangeSelection,
+} from 'roosterjs-content-model-types';
+
+const tabSpaces = '    ';
+const space = ' ';
+
+/**
+ * @internal
+ */
+export function handleTabOnParagraph(
+    model: ContentModelDocument,
+    paragraph: ContentModelParagraph,
+    rawEvent: KeyboardEvent,
+    selection: RangeSelection
+) {
+    if (paragraph.segments[0].segmentType === 'SelectionMarker' && selection.range.collapsed) {
+        setModelIndentation(model, rawEvent.shiftKey ? 'outdent' : 'indent');
+    } else {
+        const markerIndex = paragraph.segments.findIndex(
+            segment => segment.segmentType === 'SelectionMarker'
+        );
+        if (!selection.range.collapsed) {
```

Comment:
you can check collapsed by searching selected segments, if there is only one selected segment and it is SelectionMarker, that means it is collapsed.